### PR TITLE
Add initial 'weather' mod to vary cloud density, thickness, velocity

### PR DESCRIPTION
**Pull request by paramat**
screenshot_20191001_171356

screenshot_20191001_172037

Firstly, don't get too excited by the mod name.
This mod is like 'env_sounds', it is intended to be simple at first, to have more features added later as engine features improve. This is why it is called 'weather' even though it only varies cloud density, thickness and velocity.

The reason i haven't included dark clouds, rain. snow etc. yet is because those are currently problematic/intensive/network-intensive. Work on improving weather implementation is going on elsewhere (Hidden Worlds game, snowdrift mod). Considering the problems it is too early to add a complete weather system to MTG, the MTEngine may need improving first.
Clouds are very often in view, so although this is simple it makes a very significant difference to the experience.
Using my experience with the snowdrift mod (and Hidden Worlds) i will be happy to develop this mod and add weather features in future.

* This mod uses player location humidity and 4 noises to vary: density, thickness, x velocity, z velocity.
* By default, every 8sec all players have their sky updated.
* The default time scale for changes is 600sec / 10min, half of a default MT day. This keeps changes gradual but allows significant change during daylight.
* Every server session start a random time offset is added to randomise the behaviour of the noises. Without this, each server session in a world would go through identical noise variations.
* Also works with 3D clouds disabled.

Variations
------------
Density (proportion of cloud covering sky):

The initial value varies from 0.25 at humidity 0 to 1.0 at humidity 100, and is limited to those values for more extreme humidities.
Then this value is multiplied by the noise value (0.0 to 1.0).
The result is that density varies between 0.0 and 0.25 at humidity 0 (desert), and varies between 0.0 and 1.0 at humidity 100 (rainforest).
The intention being that clear sky is possible in both biomes, but cloud cover is always low in desert, but can become overcast in rainforest.

Thickness (in nodes):

The initial value varies from 8 (fairly thin clouds) at humidity 0 to 32 (twice classic thickness) at humidity 100, and is limited to those values for more extreme humidities.
Then this value is multiplied by the noise value (0.0 to 1.0).
At this stage, the value varies between 0 and 8 at humidity 0 (desert), and varies between 0 and 32 at humidity 100 (rainforest).
The value is then rounded down to an integer to make clouds a certain number of nodes thick, in the same way that their unit width is 64 nodes.
Finally a minimum of 1 node thick is enforced, so that only density variaton can make clouds disappear, not thickness variation.
The intention being that thin clouds are possible in both biomes, but clouds are always fairly thin in desert, but can become very fat in rainforest.

Velocity (in nodes per second):

Each component is varied by noise between 0 and 4 (twice classic velocity).

How to test
--------------
* In the init.lua, set TSCALE = 3 and CYCLE = 3, to cause rapidly changing noise values.
* Edit line 78 to `local humid = 0` or `local humid = 100` to see behaviour at the extremes of humidity, indoependent of what biome the player is in.

### CODE REVIEW

**Comment by Fixer-007** [comment]
I guess it is a good start, you always start with something and build up it along the way.

File: mods/weather/init.lua [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	local nobj_speedx = nobj_speedx or minetest.get_perlin(np_speedx)
+	local nobj_speedz = nobj_speedz or minetest.get_perlin(np_speedz)
+
+	local n_density = nobj_density:get2d({x = time, y = 0})
+	local n_thickness = nobj_thickness:get2d({x = time, y = 0})
+	local n_speedx = nobj_speedx:get2d({x = time, y = 0})
+	local n_speedz = nobj_speedz:get2d({x = time, y = 0})
+
+	for _, player in ipairs(minetest.get_connected_players()) do
+		local humid = minetest.get_humidity(player:getpos())
```
Comment (Panquesito7) [inline mods/weather/init.lua]: Please use `get_pos`.
```suggestion
		local humid = minetest.get_humidity(player:get_pos())
```

Comment (paramat) [reply mods/weather/init.lua]: Thanks will do.

Comment (paramat) [reply mods/weather/init.lua]: Done.

Comment (Panquesito7) [reply mods/weather/init.lua]: Thank you!

**Comment by lhofhansl** [comment]
I like it. +1

Any possibility to generate the cloud with 3D noise, so that they have variations along the Y-axis too?

**Comment by paramat** [comment]
Added code to disable the mod in mgv6 and mgsinglenode, where the Biome API doesn't function.
Added a setting to disable the mod, for people using alternative weather mods.

lhofhansl, only with engine changes.

File: mods/weather/init.lua [hunk 2]
```diff
@@ -0,0 +1,105 @@
+-- Disable by mapgen or setting
+
+local mg_name = minetest.get_mapgen_setting("mg_name")
+if mg_name == "v6" or mg_name == "singlenode" or
+		minetest.settings:get_bool("enable_weather") == false then
```
Comment (Panquesito7) [inline mods/weather/init.lua]:
```suggestion
		minetest.settings:get_bool("enable_weather") ~= true then
```

Comment (paramat) [reply mods/weather/init.lua]: This was intentional so that weather will be enabled if the setting is not defined (nil).

**Comment by DevyHeavy** [comment]
I tested this. I found that when a world is first started, or a player joins, the clouds are not updated until after CYCLE time, as expected. However, it looks weird to have the clouds suddenly change since they are not updated right away using this mod.

**Comment by paramat** [comment]
Yes good point, i should try to trigger an update on player join.
WIP

File: mods/weather/init.lua [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	local n_speedx = nobj_speedx:get2d({x = time, y = 0})
+	local n_speedz = nobj_speedz:get2d({x = time, y = 0})
+
+	for _, player in ipairs(minetest.get_connected_players()) do
+		local humid = minetest.get_humidity(player:get_pos())
+		player:set_clouds({
+			density = math.min(math.max(humid / 100, 0.25), 1.0) * n_density,
+			thickness = math.max(math.floor(
+				math.min(math.max(32 * humid / 100, 8), 32) * n_thickness
+				), 1),
```
Comment (sfan5) [inline mods/weather/init.lua]: This code is not very nice to read.
A helper `local clamp = function(lower, upper, n) return math.min(math.max(n, lower), upper)` can be used to improve this.

Comment (paramat) [reply mods/weather/init.lua]: I agree.

Comment (paramat) [reply mods/weather/init.lua]: Done.

**Pull request merged**